design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,136 @@
+to the amount of possible devices one can attach.
+
+Users of libvirt who can estimate how many potential network devices are needed,
+can define the additional PCI/e controllers in advance.
+KubeVirt does not expose such domain details through its API.
+But even if these details had been exposed by the VMI object,
+a common user would find it challenging to define such low domain details.
+
+### Goals
+- Allow users to define how many network devices can be potentially hotplug.
+- Do not expose domain architecture details (e.g. PCI/PCIe controllers).
+
+### Non-Goals
+- Define an upper limit to the number of network devices that can be defined. 
+
+## Proposal
+Users are to express the potential amount of network devices through
+the VMI spec.
+Internally, kubevirt will use libvirt logic to generate the needed
+devices to support the amount specified (e.g. PCIe controllers).
+
+Using hotplug, users may add network devices up to the amount specified
+by the new user *network-interface-list-capacity* input.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new field, named `networkInterfaceListCapacity`.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 1 and 32.
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+#### VMI Spec example
+```yaml
+
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-capacity-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+      networkInterfaceListCapacity: 5
+      rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the network-interface-list-capacity and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
```

Comment:
I would prefer this to be a follow up (added later to the design) or mentioned as optional here.
Adding this kind of validation in kubevirt is in addition to the one existing in libvirt, and we could defer this validation and depend on the underlying implementation to fail the whole thing.

While it will be nice to block immediately on the operation, in a declarative env this is not a must. Attempting to add a large validation logic with many factors and assumptions from the backend adds complexity.

Can we agree to leave it out for now and add it later?